Runners Require Nil Check

#197 connects EG components using the [watchable](https://pkg.go.dev/github.com/telepresenceio/watchable) library. Since [ProviderResources](https://github.com/envoyproxy/gateway/blob/main/internal/message/types.go#L14-L22) contain pointer slices, the elements may be nil. Runners should check for nil pointers.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x21aff6d]

goroutine 153 [running]:
github.com/envoyproxy/gateway/internal/gatewayapi/runner.(*Runner).subscribeAndTranslate(0xc000a00e00, {0x29dd170, 0xc0000bfe40})
	/Users/daneyonhansen/go/src/github.com/envoyproxy/gateway/internal/gatewayapi/runner/runner.go:65 +0x12d
created by github.com/envoyproxy/gateway/internal/gatewayapi/runner.(*Runner).Start
	/Users/daneyonhansen/go/src/github.com/envoyproxy/gateway/internal/gatewayapi/runner/runner.go:35 +0x165
```

--- a/internal/gatewayapi/runner/runner.go
+++ b/internal/gatewayapi/runner/runner.go
@@ -64,15 +64,25 @@ func (r *Runner) subscribeAndTranslate(ctx context.Context) {
 		gatewayClasses := r.ProviderResources.GetGatewayClasses()
 		// Fetch the first gateway class since there should be only 1
 		// gateway class linked to this controller
-		t := &gatewayapi.Translator{
-			GatewayClassName: v1beta1.ObjectName(gatewayClasses[0].GetName()),
+		switch {
+		case gatewayClasses == nil:
+			// No need to translate, publish empty IRs, e.g. delete operation.
+			r.XdsIR.Delete(r.Name())
+			r.InfraIR.Delete(r.Name())
+		case gatewayClasses[0] == nil:
+			panic("gatewayclass is nil")
+		default:
+			// Translate and publish IRs.
+			t := &gatewayapi.Translator{
+				GatewayClassName: v1beta1.ObjectName(gatewayClasses[0].GetName()),
+			}
+			// Translate to IR
+			result := t.Translate(&in)
+			// Publish the IRs. Use the service name as the key
+			// to ensure there is always one element in the map
+			r.XdsIR.Store(r.Name(), result.XdsIR)
+			r.InfraIR.Store(r.Name(), result.InfraIR)
 		}
-		// Translate to the IRs
-		result := t.Translate(&in)
-		// Publish the IRs. Use the service name as the key to ensure there
-		// is always one element in the map.
-		r.XdsIR.Store(r.Name(), result.XdsIR)
-		r.InfraIR.Store(r.Name(), result.InfraIR)
 	}
 	r.Logger.Info("shutting down")
 }

--- a/internal/gatewayapi/translator.go
+++ b/internal/gatewayapi/translator.go
@@ -112,6 +112,9 @@ func (t *Translator) GetRelevantGateways(gateways []*v1beta1.Gateway) []*Gateway
 	var relevant []*GatewayContext
 
 	for _, gateway := range gateways {
+		if gateway == nil {
+			panic("received nil gateway")
+		}
 		if gateway.Spec.GatewayClassName == t.GatewayClassName {
 			gc := &GatewayContext{
 				Gateway: gateway.DeepCopy(),
@@ -451,6 +454,9 @@ func (t *Translator) ProcessHTTPRoutes(httpRoutes []*v1beta1.HTTPRoute, gateways
 	var relevantHTTPRoutes []*HTTPRouteContext
 
 	for _, h := range httpRoutes {
+		if h == nil {
+			panic("received nil httproute")
+		}
 		httpRoute := &HTTPRouteContext{HTTPRoute: h}
 
 		// Find out if this route attaches to one of our Gateway's listeners,

